### PR TITLE
[QMS-1007] Fix: Endless route calculation not interruptable

### PR DESCRIPTION
**Pull request by JFritzle**
### What is the linked issue for this pull request:
[comment]: # (Add the ticket number behind QMS-# )

QMS-#1007
QMS-#388

### What you have done:
[comment]: # (Describe roughly.)

I have more or less completely adopted the changes/expansions of files 
- qmapshack/gis/rte/router/CRouterBRouter.cpp
- qmapshack/gis/rte/router/CRouterBRouter.h

from @eyzdh.

Code is now rich of comment lines which I left untouched, because mods are difficult to understand without.

During my BRouter track creation tests using new code on platforms Windows, Linux and macOS,
- I did not encounter crashes anymore 
- nor happened endless route calculation anymore.

To make sure, that others have the same experience, I would appreciate feedback for each platform from some developer/user before merging code into dev. Does track creation still work as expected?
 
### Steps to perform a simple smoke test:
[comment]: # (List the steps.)

1. Run QMS
2. Start creating track with local BRouter
3. Fire **rapidly** subsequent track points
4. Do not encounter crash or endless route calculation
5. Do not encounter anymore straight connection lines due to BRouter not responding fast enough
6. Set subsequent track points as usual
7. Verify that track creation behaves as usual

### Does the code comply to the coding rules and naming conventions [Coding Guidelines](https://github.com/Maproom/qmapshack/wiki/DeveloperCodingGuideline):

- [x] yes

### Is every user facing string in a tr() macro?

- [x] yes

### Did you add the ticket number and title into the changelog? Keep the numeric order in each release block.

- [x] yes, I didn't forget to change changelog.txt

### CODE REVIEW

**Comment by wthaem** [comment]
No issues noticed when adding track, route, or area and playing with them (adding/moving points). Longest track ~2500km, longest track segment calculation on my relatively fast CPU ~ 90sec. Cancelling calculation possible (but don't move the mouse during cancellation, otherwise the calculation restarts due to the new mouse location). No crash in the form I reported a long time ago (but here my new system can be involved, too!).

File: src/qmapshack/gis/rte/router/CRouterBRouter.h [hunk 1]
```diff
@@ -86,6 +98,59 @@ class CRouterBRouter : public IRouter, private Ui::IRouterBRouter {
...
+  // freely update pendingKey with the newest user position.
+  QTimer* timerDebounce{nullptr};
+
+  // The most-recently requested key that has not yet been sent to BRouter.
+  // Set by calcRoute() whenever a newer call arrives; cleared by
+  // startAsyncRequest() when that key is dispatched.
+  // hasPendingKey is true whenever pendingKey contains valid data.
+  bool hasPendingKey{false};
+  IGisItem::key_t pendingKey;
+
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.h:140]: This is not a good code pattern as it violates the single-source-of-truth pattern. IGisItem::key_t has a clear() method. You can add a isValid() method to test on all empty string. isValid() replaces the  hasPendingKey flag. If the pending key is processed it will be cleared.

File: src/qmapshack/gis/rte/router/CRouterBRouter.h [hunk 2]
```diff
@@ -86,6 +98,59 @@ class CRouterBRouter : public IRouter, private Ui::IRouterBRouter {
...
+  // True while the async path holds the mutex (between startAsyncRequest()
+  // and the corresponding release in slotRequestFinished / cancel).
+  bool asyncMutexHeld{false};
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.h:143]: The mutex handling in the original code was bad. With that flag it does not get better.  

As far as I understand that code we do not have a multi-threading problem here. We have another CProgressBar problem. Therefore the problem is reentrancy because of  CProgressBar operating the event loop. And not reentrancy because of two concurrent threads. 

As long as everything runs in the same thread we should never be able  to write a variable at the same time. Therefore a single flag to signal that we have a pending operation should be fine.

File: src/qmapshack/gis/rte/router/CRouterBRouter.h [hunk 3]
```diff
@@ -86,6 +98,59 @@ class CRouterBRouter : public IRouter, private Ui::IRouterBRouter {
...
+
+  // ── Route result cache ────────────────────────────────────────────────────
+  struct CachedRoute {
+    QPolygonF coords;
+    qreal costs{0.0};
+  };
+  // MAX_CACHED entries max; negligible memory, eliminates redundant requests
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.h:150]: I know there are bad examples in the code but coding guidelines require cached_route_t as structure name

https://github.com/Maproom/qmapshack/wiki/DeveloperCodingGuideline

File: src/qmapshack/gis/rte/router/CRouterBRouter.h [hunk 4]
```diff
@@ -86,6 +98,59 @@ class CRouterBRouter : public IRouter, private Ui::IRouterBRouter {
...
+  //   slotRequestFinished / cancel:
+  //     → mutex released
+  //     → scheduleNextIfPending() re-invokes calcRoute(pendingKey) if set
+  //
+  // Net effect: BRouter receives at most one request per DEBOUNCE_TIME ms idle window,
+  // the last user-set position is always eventually computed, and no stale
+  // intermediate results can overwrite a newer one.
+
+  // The in-flight reply (nullptr when idle).
+  QNetworkReply* pendingAsyncReply{nullptr};
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.h:127]: Use a QPointer. Much easier and more reliable handling of instances getting destroyed.

File: src/qmapshack/gis/rte/router/CRouterBRouter.h [hunk 5]
```diff
@@ -20,13 +20,17 @@
...
 #include <QNetworkAccessManager>
 #include <QProcess>
 #include <QTimer>
 
 #include "gis/rte/router/IRouter.h"
 #include "ui_IRouterBRouter.h"
 
+#define DEBOUNCE_TIME 200
+#define MAX_CACHED 50
+
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.h:33]: If defines are not used by the header they do not belong in the header. 

In modern c++ this:
constexpr int kDebounceTime = 200;

is better as the define has a type and the compiler can check on that type.

File: src/qmapshack/gis/rte/router/CRouterBRouter.h [hunk 6]
```diff
@@ -69,10 +73,18 @@ class CRouterBRouter : public IRouter, private Ui::IRouterBRouter {
...
   QUrl getServiceUrl() const;
+  QString buildCacheKey(const QVector<QPointF>& points, const QList<IGisItem*>& nogos) const;
+
+  // Fires the actual HTTP request for the given key.
+  // Called by the debounce timer — at this point no request is in-flight.
+  void startAsyncRequest(const IGisItem::key_t& key);
+
+  // Schedules a catch-up call for pendingKey (if set) after the mutex has
+  // been released. Used on success, error, and cancel paths alike.
+  void scheduleNextIfPending();
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.h:87]: Imho scheduleNext() would be better. Took me a couple of minutes to understand what the "If" is standing for 😆

File: src/qmapshack/gis/rte/router/CRouterBRouter.cpp [hunk 7]
```diff
@@ -76,28 +81,82 @@ CRouterBRouter::CRouterBRouter(QWidget* parent) : IRouter(false, parent) {
...
   localBRouter = new CRouterBRouterLocal(*this);
 
+  // Debounce: coalesces rapid calcRoute(key) calls into a single request.
+  // The mutex is NOT held during the debounce window — see header for the
+  // full flow description.
+  timerDebounce = new QTimer(this);
+  timerDebounce->setSingleShot(true);
+  timerDebounce->setInterval(DEBOUNCE_TIME);
+
+  routeCache.setMaxCost(MAX_CACHED);
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.cpp:91]: What is happening if the max cost is exceeded? Items are deleted immediately. Does require special handling?

File: src/qmapshack/gis/rte/router/CRouterBRouter.cpp [hunk 8]
```diff
@@ -76,28 +81,82 @@ CRouterBRouter::CRouterBRouter(QWidget* parent) : IRouter(false, parent) {
...
+// ---------------------------------------------------------------------------
+// Destructor
+// ---------------------------------------------------------------------------
 CRouterBRouter::~CRouterBRouter() {
   isShutdown = true;
+
+  timerDebounce->stop();
+  hasPendingKey = false;
+
+  if (pendingAsyncReply) {
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.cpp:111]: ok, but it's nicer to compare with nullptr.

File: src/qmapshack/gis/rte/router/CRouterBRouter.cpp [hunk 9]
```diff
@@ -106,6 +165,7 @@ void CRouterBRouter::slotToolSetupClicked() {
   setup->load();
   getBRouterVersion();
   updateDialog();
+  routeCache.clear();
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.cpp:168]: Why at this point. And what about all the other async stuff?

File: src/qmapshack/gis/rte/router/CRouterBRouter.cpp [hunk 10]
```diff
@@ -232,26 +292,25 @@ QNetworkRequest CRouterBRouter::getRequest(const QVector<QPointF>& routePoints,
         QPolygonF polygon;
         line->getPolylineDegFromData(polygon);
         QString nogoPoints;
-        for (const QPointF point : std::as_const(polygon)) {
+        for (const QPointF& point : std::as_const(polygon)) {
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.cpp:295]: using a const reference is perfect 👍 And in this case we can get rid of std::as_const.

File: src/qmapshack/gis/rte/router/CRouterBRouter.cpp [hunk 11]
```diff
@@ -290,15 +365,27 @@ int CRouterBRouter::calcRoute(const QPointF& p1, const QPointF& p2, QPolygonF& c
   QList<IGisItem*> nogos;
   CGisWorkspace::self().getNogoAreas(nogos);
 
+  const QString cacheKey = buildCacheKey(points, nogos);
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.cpp:368]: const ref

File: src/qmapshack/gis/rte/router/CRouterBRouter.cpp [hunk 12]
```diff
@@ -290,15 +365,27 @@ int CRouterBRouter::calcRoute(const QPointF& p1, const QPointF& p2, QPolygonF& c
...
+  if (const CachedRoute* cached = routeCache.object(cacheKey)) {
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.cpp:369]: imho assigning cached first and then checking for nullptr in the if is a bit nicer.

File: src/qmapshack/gis/rte/router/CRouterBRouter.cpp [hunk 13]
```diff
@@ -307,145 +394,279 @@ int CRouterBRouter::synchronousRequest(const QVector<QPointF>& points, const QLi
   synchronous = true;
 
   QNetworkReply* reply = networkAccessManager->get(getRequest(points, nogos));
+  auto replyGuard = qScopeGuard([reply] { reply->deleteLater(); });
+  Q_UNUSED(replyGuard);
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.cpp:398]: What for? And if, why not for mutexGuard?

**Review by kiozen — COMMENTED** [review]
I am not done yet with the PR. But I have to switch task. Will continue later. Don't bother about the review yet.

**Comment by JFritzle** [comment]
I am sorry, program logic is beyond my understanding. I have to delegate code back to @eyzdh and wait for an update.

**Comment by wthaem** [comment]
**Task: Given the waypoints of the attached GPX files, try to build a route from P1 to P7 using BRouter. Choose trekking option.**

[broutertest1.zip](https://github.com/user-attachments/files/26059981/broutertest1.zip)


The calculation of each leg/route segment takes in my case 10 sec. or more due to the large distance between the points.

* Go to P1 and select "add route" from the context menu at this point
* Try to click rapidly one after the other the given waypoints with the left mouse button and look what happens
* Notice the erratic/irritating behaviour of the last shown green square depending on the mouse position: either the last click gets lost, or a new calculation starts at some intermediate and unwanted new mouse position, or clicking the mouse at the next point is not possible as long as the calculation runs, ...

This behaviour can't be seen with short distances between the waypoints.

More observations:

*Info window isn't removed:*

* Routing engine switched from BRouter to Routino
* Define new route from waypoints - triggers automatic route recalculation
* GUI shows still BRouter info window - no chance to remove it, the window reappears even after loading WMTS map tiles!

BRouterRequest


*Jumping roads:*

* Used map: velo DE (gmapsupp type)
* A waypoint (P1) seems to be exactly positioned on a road
* Depending on the zoom level, it is either on the road or only near the road as shown in the attached snapshots
* Doesn't happen with the OSM Topo map (WMTS type). P1 is exactly on the road for each zoom level.
* My conclusion: not the waypoint but the underlying road location depends on the zoom level (can be an understandable coordinate rounding problem)
* This is important for the construction of routes or tracks because routing depends on the (nearly) exact position of a point on a road. If the road is at a wrong position, the selected point will not be on the road (the one used by the routing engine) and the routing engine may fail.

capture_20260317_160933
capture_20260317_160849
capture_20260317_160822

File: src/qmapshack/gis/rte/router/CRouterBRouter.cpp [hunk 14]
```diff
@@ -307,145 +394,279 @@ int CRouterBRouter::synchronousRequest(const QVector<QPointF>& points, const QLi
...
-    const QByteArray& res = reply->readAll();
+  if (res.isEmpty()) {
+    throw tr("response is empty");
+  }
 
-    if (res.isEmpty()) {
-      throw tr("response is empty");
+  QDomDocument xml;
+  {
+    const QDomDocument::ParseResult parseResult = xml.setContent(res);
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.cpp:442]: for complex objects always const &

File: src/qmapshack/gis/rte/router/CRouterBRouter.cpp [hunk 15]
```diff
@@ -307,145 +394,279 @@ int CRouterBRouter::synchronousRequest(const QVector<QPointF>& points, const QLi
...
+  // Do NOT call rte->reset() here — the old route stays visible while the
+  // new one is being calculated, which avoids the visual flicker of having
+  // nothing displayed during the request. reset() is called in
+  // slotRequestFinished() just before applying the new result.
+  slotCloseStatusMsg();
+
   synchronous = false;
 
   QNetworkReply* reply = networkAccessManager->get(getRequest(points, nogos));
+  pendingAsyncReply = reply;
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.cpp:585]: This code is really hard to read and check. You do stuff in one method and assume that stuff is done in another. Are you sure pendingAsyncReply is always nullptr? If not we have a memory leak.  And it's impossible for someone else to see all these cross method dependencies. The recipe for introducing regressions with future changes. 

Use at least a smart pointer (std::unique_ptr) to prevent memory leaks. In C++ you have to take care of pointers. You have to define who owns the pointer. Who observes the pointer. The owner instantiates the pointer and deletes it. The observer has to notice when a pointer is deleted.  All this should be at very defined single locations. Else disaster is pre-ordered.

Anyway, this is really just bitching around another problem caused by CProgressBar. Imho a threaded approach would be much cleaner. By that everything is calculated in a single thread and when done the UI is updated.

File: src/qmapshack/gis/rte/router/CRouterBRouter.cpp [hunk 16]
```diff
@@ -307,145 +394,279 @@ int CRouterBRouter::synchronousRequest(const QVector<QPointF>& points, const QLi
...
+  // Report "calculating" status — do NOT trigger a full canvas repaint here.
+  // The canvas will redraw itself when slotRequestFinished() applies the new
+  // route, avoiding one spurious repaint per request.
   CCanvas* canvas = CMainWindow::self().getVisibleCanvas();
   if (canvas) {
-    canvas->slotTriggerCompleteUpdate(CCanvas::eRedrawGis);
     canvas->reportStatus("BRouter", tr("<b>BRouter</b><br/>Routing request sent to server. Please wait..."));
   }
 
+  delete progress;
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.cpp:601]: again one of the smart pointers, either Qt or std would make this much safer

File: src/qmapshack/gis/rte/router/CRouterBRouter.cpp [hunk 17]
```diff
@@ -307,145 +394,279 @@ int CRouterBRouter::synchronousRequest(const QVector<QPointF>& points, const QLi
...
+
+  // Release the mutex and schedule the next pending key on every exit path.
+  auto cleanupGuard = qScopeGuard([this] {
+    if (asyncMutexHeld) {
+      asyncMutexHeld = false;
+      mutex.unlock();
+    }
+    scheduleNextIfPending();
+  });
+  Q_UNUSED(cleanupGuard);
```
Comment (kiozen) [inline src/qmapshack/gis/rte/router/CRouterBRouter.cpp:650]: one should be enough

**Review by kiozen — CHANGES_REQUESTED** [review]
Well the patch solves a couple of problems for sure. But it introduces a lot of complexity, too. Without diving deep into that code  I can't tell if there is a more elegant way to solve it as a single threaded / reentrant by event loop approach. Or if it is easier to solve by a thread. What I can tell for sure: The progress bar has to go. To get rid of CProgressBar is top priority as it is the root of many problems. 

It would be really nice to rethink the code to make it  easier instead of making it more complex.  Especially this cross method handling of data is painfully known to me. It never ended well in the past.

**Comment by JFritzle** [comment]
@wthaem - about _jumping roads_:

I would overlay a semitransparent OSM Topo map over velo DE map. Roads _should_ be exactly on top of each other. Do they at all for any zoom level? If point p1 deviates from velo DE map's road when zooming but stays on  OSM Topo map's road, then I guess complete velo DE map's road does no longer cover OSM Topo map's road.
-> In my opinion, this is the only possible explanation: Either velo DE map data get inaccurate when zooming or projection of map data onto map canvas gets inaccurate.

**Comment by wthaem** [comment]
@JFritzle: You are almost correct. There is yet another parameter influencing the accuracy (I didn't expect this!): the details level selected for the vector map as you can see from the attached snapshots (same zoom level, different level of detail, snapshots show overlay of 2 maps!).

capture_20260317_195650
capture_20260317_195521

**Comment by JFritzle** [comment]
@wthaem 

I took left screenshot and overlayed semitransparent right screenshot
applied with a slight clockwise twist of about 6 degrees :
grafik
What do we see: roads again exactly match. Road's slight bend to the right near to point p1 fits exactly!
-> Conclusion: changing velo DE map's level of detail applies a twist on projection. Maybe inaccuracy in computing projection?

But this seems to be an issue different from QMS-#1007,

**Comment by wthaem** [comment]
@JFritzle : 

> But this seems to be an issue different from QMS-https://github.com/Maproom/qmapshack/issues/1007,

Sure, I came across this one when preparing the example for the BRouter test and was confused that I did not hit the road with my point.

And: it is no projection issue. It is the (understandable) reduction of the number of points used to draw the road (using DP algorithm or something similar). Load the velo map, look at the Bundesstraße (especially in the nearby location of Haselünne) involved at different detail levels, and see what happens.

For me, it is a feature, not an issue!

**Comment by kiozen** [comment]
Garmin maps have layers. For outer zoom levels and more closer zoom levels. The details factor is manipulating the decision about what layer to use at what zoom level. Each layer uses a different number of bits to code coordinates. And it has a different number of elements like roads, and waypoints. 

This is fine as long as the layer is displayed at a matching zoom level. However if a layer with less resolution is displayed at a closer zoom level, the quantification errors become very obvious. At closest zoom even the error of the 24bit resolution is obvious, e.g.  badly shaped buildings. 

This is also a problem if you use "follow polyline" while drawing tracks. If you do that on an outer zoomlevel and zoom in afterwards the line is not matching the line of the map.

**Comment by JFritzle** [comment]
@wthaem 
I downloaded and installed velomap_germany_14.03.2026.
This is what Garmin BaseCamp shows from left to right - Lowest level of detail - Highest level of detail - Overlay of both:
grafik
-> Roads match exactly. No twist. Independent from zoom level.

Then I created a gmapsupp for QMS -> Roads match exactly for each level of detail too. Independent from zoom level.
And: Roads match exactly with OSM Topo map.

But there are numerous options either for installing downloaded map or for creating gmapsupp later.
Propably these options influence accuracy of levels of detail?

**Comment by wthaem** [comment]
@JFritzle : see kiozen's comment. It explains everything. And we discuss QMS, not Basecamp.

**Comment by JFritzle** [comment]
@wthaem: Sure, we discuss QMS. See my comment: At my site, no inaccuracy depending on level of detail nor zoom level.

**Comment by JFritzle** [comment]
As I have not been able to provide a own suitable solution, I will no longer work on issue.
My PR gets closed to allow someone else to contribute his solution.
